src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.cs

```diff
@@ -525,6 +525,135 @@ public override BoundNode VisitTypeOrInstanceInitializers(BoundTypeOrInstanceIni
             return new BoundStatementList(node.Syntax, rewrittenStatements, node.HasErrors);
         }
 
+        public override BoundNode VisitArrayAccess(BoundArrayAccess node)
+        {
+            // If the array access index is of type System.Index or System.Range
+            // we need to translate it into a call into an actual indexer, instead
+            // of a simple array element access.
+
+            if (this.EmitModule is null || 
+                node.Indices.Length != 1 ||
+                node.Indices[0].Type.SpecialType != SpecialType.None)
+            {
+                return base.VisitArrayAccess(node);
+            }
+
+            // PROTOTYPE: THIS MUST BE REMOVED BEFORE DEV16
+            // Since it takes a long time to update mscorlib we fake up the right
+            // code to pretend there are two indexers that support Index and Range
+            // on arrays.
+
+            var syntax = node.Syntax;
+            var F = _factory;
+            var indexLocal = F.StoreToTemp(
+                VisitExpression(node.Indices[0]),
+                out BoundAssignmentOperator indexAssign);
+            var arrayLocal = F.StoreToTemp(
+                VisitExpression(node.Expression),
+                out BoundAssignmentOperator arrayAssign);
+            var indexType = VisitType(node.Indices[0].Type);
```

Comment:
Why use VisitType here vs. just using the type of `indexLocal`?

Reply:
I can't think of a situation it could matter right now, but it's usually standard practice to visit each element individually, rather than rely on things being the same. Sometimes there are operations where the types are actually different but compatible (e.g. tuple names) so I don't want to drop things on the floor.